pkg/monitoring/services/services.go

```diff
@@ -0,0 +1,290 @@
+		)
+	}
+
+	return services, scanner.Err()
+}
+
+var sysVinitServiceRE = regexp.MustCompile(`^\s+\[\s+([\+\-\?]])\s+\]\s+(.*)$`)
+
+// ListSysVinitServices list SysVinit services via `service --status-all`
+func ListSysVinitServices() ([]SysVService, error) {
+	cmd := exec.Command("service",
+		"--status-all",
+	)
+
+	setPathEnvVar(cmd)
+	var outb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &outb
+	err := cmd.Run()
+	if err != nil {
+		errOutput, _ := ioutil.ReadAll(&outb)
+		return nil, fmt.Errorf("service: %s, %s", err.Error(), string(errOutput))
+	}
+
+	var services []SysVService
+	scanner := bufio.NewScanner(&outb)
+
+	for scanner.Scan() {
+		parts := sysVinitServiceRE.FindStringSubmatch(scanner.Text())
+
+		var state string
+
+		switch parts[1] {
+		case "+":
+			state = "running"
+		case "-":
+			state = "stopped"
+		case "?":
+			state = "unknown"
+		}
+
+		services = append(services,
+			SysVService{UnitFile: parts[2], Status: state},
+		)
+	}
+
+	return services, scanner.Err()
+}
+
+// ListUpstartServices list upstart services via `initctl list`. Returns []SysVService because Upstart is compatible with SysVInit and has the same details
+func ListUpstartServices() ([]SysVService, error) {
+	cmd := exec.Command("initctl",
+		"list",
+	)
+	setPathEnvVar(cmd)
+
+	var outb bytes.Buffer
+	cmd.Stdout = &outb
+	cmd.Stderr = &outb
+	err := cmd.Run()
+	if err != nil {
+		errOutput, _ := ioutil.ReadAll(&outb)
+		return nil, fmt.Errorf("Initctl: %s, %s", err.Error(), string(errOutput))
+	}
+
+	var services []SysVService
+	scanner := bufio.NewScanner(&outb)
+	for scanner.Scan() {
+		parts := strings.Fields(scanner.Text())
+
+		if len(parts) < 2 {
+			continue
+		}
+
+		// filter network interfaces from results e.g.
+		// network-interface-security (network-interface/ip6_vti0) start/running
+		// network-interface (dummy0) start/running
+		if strings.HasPrefix(parts[0], "network-interface") {
+			continue
+		}
+
+		stateParts := strings.Split(parts[1], "/")
+
+		services = append(services,
+			SysVService{UnitFile: parts[0], Status: stateParts[1]},
+		)
+	}
+
+	return services, scanner.Err()
+}
+
+func isSystemd() bool {
+	if _, err := os.Stat("/run/systemd/system"); err == nil {
+		return true
+	}
+	return false
+}
+
+func isUpstart() bool {
+	if _, err := os.Stat("/sbin/upstart-udev-bridge"); err == nil {
+		return true
+	}
+	cmd := exec.Command("initctl", "--version")
+	setPathEnvVar(cmd)
+	if out, err := cmd.Output(); err == nil {
+		if strings.Contains(string(out), "initctl (upstart") {
+			return true
+		}
+	}
+
+	return false
+}
+
+func setPathEnvVar(cmd *exec.Cmd) {
+	// path PATH env var to be able to found right exec location
+	cmd.Env = append(cmd.Env, "PATH="+os.Getenv("PATH"))
+}
+
+// ListServices detect the linux system manager and parse/combine results
+func ListServices() (map[string]interface{}, error) {
```

Comment:
The function is not super dramatic but I think it can be done a bit cleaner.

I suggest t to put the actual fetching for each service in its own function.
Determine which init system we're running on and then do a `switch` to call the corresponding sub-function.

Or does this not work for some reason?

Reply:
>I suggest t to put the actual fetching for each service in its own function.

this is already done because there are `ListSystemdServices()`, `ListSysVinitServices()`, `ListUpstartServices()`
- `ListServices()`   reads, transform to map and combine(sysVinit+upstart) the different managers' services list
- systemd has different struct type (and output)
- it uses the map to combine the results(of sysVinit and upstart) by the service name

Maybe I can rework it so `ListSystemdServices()`, `ListSysVinitServices()`, `ListUpstartServices()` will return `map[string]string` instead of struct. But I don't like this because it becomes not clear when you can't fully describe what the func is excepted to return